File: app/schemas/report.py
from datetime import datetime

from pydantic import BaseModel


class SalesSummary(BaseModel):
    gross_revenue: float
    total_discounts: float
    total_revenue: float
    total_refunds: float
    net_revenue: float
    order_count: int
    average_order_value: float
    # Per-sale COGS from the order_items.unit_cost snapshot (migration
    # 0020). cogs_known_ratio < 1 means some sold lines had no cost
    # snapshot, so the margin is computed over known-cost lines only.
    cogs_total: float = 0.0
    gross_margin_amount: float = 0.0
    gross_margin_percent: float | None = None
    cogs_known_ratio: float | None = None


class PaymentBreakdownItem(BaseModel):
    payment_method: str
    count: int
    amount: float


class ShiftReport(BaseModel):
    reconciliation_id: int
    drawer_session_id: int
    opened_at: datetime | None
    closed_at: datetime | None
    operator_name: str | None
    closed_by_name: str | None
    starting_cash: float
    expected_cash: float
    counted_cash: float
    cash_variance: float
    expected_non_cash: float
    counted_non_cash: float
    non_cash_variance: float
    cash_sales_total: float
    non_cash_sales_total: float
    refunds_total: float
    gross_sales_total: float
    net_sales_total: float
    completed_order_count: int
    payment_breakdown: list[PaymentBreakdownItem] = []


class DailyShiftItem(BaseModel):
    reconciliation_id: int
    drawer_session_id: int
    opened_at: datetime | None
    closed_at: datetime | None
    operator_name: str | None
    cash_sales_total: float
    non_cash_sales_total: float
    refunds_total: float
    gross_sales_total: float
    net_sales_total: float
    completed_order_count: int
    cash_variance: float


class DailyCloseTotals(BaseModel):
    gross_sales_total: float
    net_sales_total: float
    cash_sales_total: float
    non_cash_sales_total: float
    refunds_total: float
    cash_variance: float
    non_cash_variance: float
    completed_order_count: int
    shift_count: int


class DailyClose(BaseModel):
    date: str
    totals: DailyCloseTotals
    shifts: list[DailyShiftItem]


class TopProductItem(BaseModel):
    product_id: int
    product_name: str
    product_sku: str
    total_quantity_sold: int
    total_revenue: float


class TopCustomerItem(BaseModel):
    customer_id: int
    customer_name: str
    customer_email: str | None
    order_count: int
    total_spent: float
    points_balance: int


class CategorySalesItem(BaseModel):
    category_id: int | None
    category_name: str
    total_revenue: float
    total_quantity_sold: int


class SupplierSpendItem(BaseModel):
    supplier_id: int
    supplier_name: str
    po_count: int
    invoice_count: int
    approved_total: float
    variance_total: float


class SupplierSpendSummary(BaseModel):
    start_date: datetime | None = None
    end_date: datetime | None = None
    cogs_estimate: float = 0.0
    items: list[SupplierSpendItem] = []


class VarianceTrendItem(BaseModel):
    period: str  # YYYY-MM
    invoice_count: int
    billed_total: float
    approved_total: float
    variance_total: float


class VarianceTrendSummary(BaseModel):
    start_date: datetime | None = None
    end_date: datetime | None = None
    months: list[VarianceTrendItem] = []

File: client/lib/core/models/report.dart
library;

class TopProductItem {
  final int productId;
  final String productName;
  final String productSku;
  final int totalQuantitySold;
  final double totalRevenue;

  const TopProductItem({
    required this.productId,
    required this.productName,
    required this.productSku,
    required this.totalQuantitySold,
    required this.totalRevenue,
  });

  factory TopProductItem.fromJson(Map<String, dynamic> json) => TopProductItem(
    productId: json['product_id'] as int,
    productName: json['product_name'] as String,
    productSku: json['product_sku'] as String,
    totalQuantitySold: json['total_quantity_sold'] as int? ?? 0,
    totalRevenue: (json['total_revenue'] as num?)?.toDouble() ?? 0,
  );
}

class CategorySalesItem {
  final int categoryId;
  final String categoryName;
  final double totalRevenue;
  final int totalQuantitySold;

  const CategorySalesItem({
    required this.categoryId,
    required this.categoryName,
    required this.totalRevenue,
    required this.totalQuantitySold,
  });

  factory CategorySalesItem.fromJson(Map<String, dynamic> json) =>
      CategorySalesItem(
        categoryId: json['category_id'] as int,
        categoryName: json['category_name'] as String,
        totalRevenue: (json['total_revenue'] as num?)?.toDouble() ?? 0,
        totalQuantitySold: json['total_quantity_sold'] as int? ?? 0,
      );
}

class SalesSummary {
  final double grossRevenue;
  final double totalDiscounts;
  final double totalRevenue;
  final double totalRefunds;
  final double netRevenue;
  final int orderCount;
  final double averageOrderValue;

  /// Per-sale COGS from the order_items.unit_cost snapshot. Lines sold
  /// before cost tracking carry no snapshot and are excluded, so
  /// [cogsKnownRatio] < 1 signals partial coverage.
  final double cogsTotal;
  final double grossMarginAmount;
  final double? grossMarginPercent;
  final double? cogsKnownRatio;

  const SalesSummary({
    required this.grossRevenue,
    required this.totalDiscounts,
    required this.totalRevenue,
    required this.totalRefunds,
    required this.netRevenue,
    required this.orderCount,
    required this.averageOrderValue,
    this.cogsTotal = 0,
    this.grossMarginAmount = 0,
    this.grossMarginPercent,
    this.cogsKnownRatio,
  });

  factory SalesSummary.fromJson(Map<String, dynamic> json) => SalesSummary(
    grossRevenue: (json['gross_revenue'] as num?)?.toDouble() ?? 0,
    totalDiscounts: (json['total_discounts'] as num?)?.toDouble() ?? 0,
    totalRevenue: (json['total_revenue'] as num?)?.toDouble() ?? 0,
    totalRefunds: (json['total_refunds'] as num?)?.toDouble() ?? 0,
    netRevenue: (json['net_revenue'] as num?)?.toDouble() ?? 0,
    orderCount: json['order_count'] as int? ?? 0,
    averageOrderValue: (json['average_order_value'] as num?)?.toDouble() ?? 0,
    cogsTotal: (json['cogs_total'] as num?)?.toDouble() ?? 0,
    grossMarginAmount:
        (json['gross_margin_amount'] as num?)?.toDouble() ?? 0,
    grossMarginPercent: (json['gross_margin_percent'] as num?)?.toDouble(),
    cogsKnownRatio: (json['cogs_known_ratio'] as num?)?.toDouble(),
  );
}
